scalameta/semantic/shared/src/main/scala/scala/meta/internal/semantic/Meta.scala

```diff
@@ -0,0 +1,82 @@
+package scala.meta
+package internal
+package semantic
+
+import java.nio.charset.Charset
+import org.scalameta.data._
+import scala.{Seq => _}
+import scala.collection.immutable.Seq
+import scala.meta.io._
+import scala.meta.internal.semantic.{schema => s}
+import scala.meta.{semantic => m}
+import scala.meta.{Dialect => mDialect}
+import scala.meta.inputs.{Input => mInput, Position => mPosition, Point => mPoint}
+
+package object meta {
+  implicit class XtensionDatabaseSchema(mdatabase: m.Database) {
+    def toSchema(sourcepath: Sourcepath): s.Database = {
+      val sentries = mdatabase.entries.map {
+        case (minput, m.Attributes(mdialect, mnames, mmessages, mdenots, msugars)) =>
+          object mDialect {
+            def unapply(mdialect: mDialect): Option[String] = {
+              val isStandard = scala.meta.Dialect.standards.exists(_._2 == mdialect)
+              if (isStandard) Some(mdialect.toString) else None
+            }
+          }
+          object mPosition {
+            def unapply(mpos: mPosition): Option[s.Range] = mpos match {
+              case scala.meta.inputs.Position.Range(`minput`, mPoint.Offset(_, sstart), mPoint.Offset(_, send)) => Some(s.Range(sstart, send))
+              case _ => None
+            }
+          }
+          object mSymbol {
+            def unapply(msym: m.Symbol): Option[String] = Some(msym.syntax)
+          }
+          object mSeverity {
+            def unapply(msev: m.Severity): Option[s.Message.Severity] = {
+              msev match {
+                case m.Severity.Info => Some(s.Message.Severity.INFO)
+                case m.Severity.Warning => Some(s.Message.Severity.WARNING)
+                case m.Severity.Error => Some(s.Message.Severity.ERROR)
+                case _ => None
+              }
+            }
+          }
+          object mDenotation {
+            def unapply(mdenot: m.Denotation): Option[s.Denotation] = mdenot match {
+              case m.Denotation(sflags, sname, sinfo) => Some(s.Denotation(sflags, sname, sinfo))
+              case _ => None
+            }
+          }
+          val spath = minput match {
+            case mInput.File(path, charset) if charset == Charset.forName("UTF-8") => sourcepath.relativize(path.toURI).getOrElse(sys.error(s"can't find $path in $sourcepath"))
+            case other => sys.error(s"bad database: unsupported input $other")
+          }
+          val sdialect = mdialect match {
+            case mDialect(sdialect) => sdialect
```

Comment:
Could we do something like `m.Dialect.standards.find(_.name == mdialect).getOrElse(sys.error)`? This whole extractor business is not helping readability IMO.